In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set path

path = r'C:\Users\Josh Wattay\OneDrive\Documents\Desktop\Flight Delay Analysis - Raw Data\2019'

In [3]:
# Import Datasets

df = pd.read_csv(os.path.join(path, 'flights2019.csv'), index_col = False)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_11716\1827886079.py:3: DtypeWarning: Columns (11,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path, 'flights2019.csv'), index_col = False)


In [4]:
#Shape check
df.shape

(8091684, 70)

In [5]:
# Set the display option to show all 70 columns
pd.set_option('display.max_columns', 70)

In [6]:
#Inspect Dataframe and begin Cleaning by ensuring columns are named appropriately
df.head()

Unnamed: 0  Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0           0  2019        1      1           4          5  2019-01-04   
1           1  2019        1      1           4          5  2019-01-04   
2           2  2019        1      1           4          5  2019-01-04   
3           3  2019        1      1           4          5  2019-01-04   
4           4  2019        1      1           4          5  2019-01-04   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        UA                     19977   
1                        UA                     19977   
2                        UA                     19977   
3                        UA                     19977   
4                        UA                     19977   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          UA                             5657   
1                          UA                             5658   
2                          UA                             5658   
3                          UA                             5659   
4                          UA                             5660   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 OO   
1                                               NaN                 OO   
2                                               NaN                 OO   
3                                               NaN                 OO   
4                                               NaN                 OO   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20304                          OO      N945SW   
1                     20304                          OO      N932SW   
2                     20304                          OO      N932SW   
3                     20304                          OO      N916SW   
4                     20304                          OO      N107SY   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             5657            14698             1469802   
1                             5658            12266             1226603   
2                             5658            14783             1478302   
3                             5659            12389             1238902   
4                             5660            13342             1334207   

   OriginCityMarketID Origin       OriginCityName OriginState  \
0               34236    SBP  San Luis Obispo, CA          CA   
1               31453    IAH          Houston, TX          TX   
2               34783    SGF      Springfield, MO          MO   
3               32389    ISN        Williston, ND          ND   
4               33342    MKE        Milwaukee, WI          WI   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0                6      California          14771           1477104   
1               48           Texas          15919           1591904   
2               29        Missouri          12266           1226603   
3               38    North Dakota          11292           1129202   
4               55       Wisconsin          11292           1129202   

   DestCityMarketID Dest       DestCityName DestState  DestStateFips  \
0             32457  SFO  San Francisco, CA        CA              6   
1             31834  XNA   Fayetteville, AR        AR              5   
2             31453  IAH        Houston, TX        TX             48   
3             30325  DEN         Denver, CO        CO              8   
4             30325  DEN         Denver, CO        CO              8   

  DestStateName  CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0    California        1400   1353.0      -7.0              0.0       0.0   
1      Arkansas         935    930.0      -5.0              0.0       0.0   
2         Tex

The 'Unnamed: 0' column will be removed as it is an additional index and redundant. 

In [7]:
# Drop the 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

In [8]:
df.shape

(8091684, 69)

In [9]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0  2019        1      1           4          5  2019-01-04   
1  2019        1      1           4          5  2019-01-04   
2  2019        1      1           4          5  2019-01-04   
3  2019        1      1           4          5  2019-01-04   
4  2019        1      1           4          5  2019-01-04   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        UA                     19977   
1                        UA                     19977   
2                        UA                     19977   
3                        UA                     19977   
4                        UA                     19977   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          UA                             5657   
1                          UA                             5658   
2                          UA                             5658   
3                          UA                             5659   
4                          UA                             5660   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 OO   
1                                               NaN                 OO   
2                                               NaN                 OO   
3                                               NaN                 OO   
4                                               NaN                 OO   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20304                          OO      N945SW   
1                     20304                          OO      N932SW   
2                     20304                          OO      N932SW   
3                     20304                          OO      N916SW   
4                     20304                          OO      N107SY   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             5657            14698             1469802   
1                             5658            12266             1226603   
2                             5658            14783             1478302   
3                             5659            12389             1238902   
4                             5660            13342             1334207   

   OriginCityMarketID Origin       OriginCityName OriginState  \
0               34236    SBP  San Luis Obispo, CA          CA   
1               31453    IAH          Houston, TX          TX   
2               34783    SGF      Springfield, MO          MO   
3               32389    ISN        Williston, ND          ND   
4               33342    MKE        Milwaukee, WI          WI   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0                6      California          14771           1477104   
1               48           Texas          15919           1591904   
2               29        Missouri          12266           1226603   
3               38    North Dakota          11292           1129202   
4               55       Wisconsin          11292           1129202   

   DestCityMarketID Dest       DestCityName DestState  DestStateFips  \
0             32457  SFO  San Francisco, CA        CA              6   
1             31834  XNA   Fayetteville, AR        AR              5   
2             31453  IAH        Houston, TX        TX             48   
3             30325  DEN         Denver, CO        CO              8   
4             30325  DEN         Denver, CO        CO              8   

  DestStateName  CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0    California        1400   1353.0      -7.0              0.0       0.0   
1      Arkansas         935    930.0      -5.0              0.0       0.0   
2         Texas         643    637.0      -6.0              0.0       0.0   
3      C

Rename syntax for various columns for consistency and accuracy.

In [10]:
# Define the dictionary for renaming columns
rename_dict = {
    'DayofMonth': 'Day_of_Month',
    'DayOfWeek': 'Day_of_Week',
    'FlightDate': 'Flight_Date',
    'Operating_Airline': 'Operating Airline',
    'OriginAirportID': 'Origin_Airport_ID',
    'OriginAirportSeqID': 'Origin_Airport_Seq_ID',
    'OriginCityMarketID': 'Origin_City_Market_ID',
    'OriginCityName': 'Origin_City_Name',
    'OriginState': 'Origin_State',
    'OriginStateFips': 'Origin_State_Fips',
    'OriginStateName': 'Origin_State_Name',
    'DestAirportID': 'Dest_Airport_ID',
    'DestAirportSeqID': 'Dest_Airport_Seq_ID',
    'DestCityMarketID': 'Dest_City_Market_ID',
    'DestCityName': 'Dest_City_Name',
    'DestState': 'Dest_State',
    'DestStateFips': 'Dest_State_Fips',
    'DestStateName': 'Dest_State_Name',
    'CRSDepTime': 'CRS_Dep_Time',
    'DepTime': 'Dep_Time',
    'DepDelay': 'Dep_Delay',
    'DepDelayMinutes': 'Dep_Delay_Minutes',
    'DepDel15': 'Dep_Del_15',
    'DepartureDelayGroups': 'Departure_Delay_Groups',
    'DepTimeBlk': 'Dep_Time_Blk',
    'TaxiOut': 'Taxi_Out',
    'WheelsOff': 'Wheels_Off',
    'WheelsOn': 'Wheels_On',
    'TaxiIn': 'Taxi_In',
    'CRSArrTime': 'CRS_Arr_Time',
    'ArrTime': 'Arr_Time',
    'ArrDelay': 'Arr_Delay',
    'ArrDelayMinutes': 'Arr_Delay_Minutes',
    'ArrDel15': 'Arr_Del_15',
    'ArrivalDelayGroups': 'Arrival_Delay_Groups',
    'ArrTimeBlk': 'Arr_Time_Blk',
    'CancellationCode': 'Cancellation_Code',
    'CRSElapsedTime': 'CRS_Elapsed_Time',
    'ActualElapsedTime': 'Actual_Elapsed_Time',
    'AirTime': 'Air_Time',
    'DistanceGroup': 'Distance_Group',
    'CarrierDelay': 'Carrier_Delay',
    'WeatherDelay': 'Weather_Delay',
    'NASDelay': 'NAS_Delay',
    'SecurityDelay': 'Security_Delay',
    'LateAircraftDelay': 'Late_Aircraft_Delay',
    'FirstDepTime': 'First_Dep_Time',
    'TotalAddGTime': 'Total_Add_G_Time',
    'LongestAddGTime': 'Longest_Add_G_Time',
    'DivDistance': 'Div_Distance'
}

# Rename the columns
df = df.rename(columns=rename_dict)

In [11]:
df.shape

(8091684, 69)

In [12]:
df.head()

Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2019        1      1             4            5  2019-01-04   
1  2019        1      1             4            5  2019-01-04   
2  2019        1      1             4            5  2019-01-04   
3  2019        1      1             4            5  2019-01-04   
4  2019        1      1             4            5  2019-01-04   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        UA                     19977   
1                        UA                     19977   
2                        UA                     19977   
3                        UA                     19977   
4                        UA                     19977   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          UA                             5657   
1                          UA                             5658   
2                          UA                             5658   
3                          UA                             5659   
4                          UA                             5660   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 OO   
1                                               NaN                 OO   
2                                               NaN                 OO   
3                                               NaN                 OO   
4                                               NaN                 OO   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20304                          OO      N945SW   
1                     20304                          OO      N932SW   
2                     20304                          OO      N932SW   
3                     20304                          OO      N916SW   
4                     20304                          OO      N107SY   

   Flight_Number_Operating_Airline  Origin_Airport_ID  Origin_Airport_Seq_ID  \
0                             5657              14698                1469802   
1                             5658              12266                1226603   
2                             5658              14783                1478302   
3                             5659              12389                1238902   
4                             5660              13342                1334207   

   Origin_City_Market_ID Origin     Origin_City_Name Origin_State  \
0                  34236    SBP  San Luis Obispo, CA           CA   
1                  31453    IAH          Houston, TX           TX   
2                  34783    SGF      Springfield, MO           MO   
3                  32389    ISN        Williston, ND           ND   
4                  33342    MKE        Milwaukee, WI           WI   

   Origin_State_Fips Origin_State_Name  Dest_Airport_ID  Dest_Airport_Seq_ID  \
0                  6        California            14771              1477104   
1                 48             Texas            15919              1591904   
2                 29          Missouri            12266              1226603   
3                 38      North Dakota            11292              1129202   
4                 55         Wisconsin            11292              1129202   

   Dest_City_Market_ID Dest     Dest_City_Name Dest_State  Dest_State_Fips  \
0                32457  SFO  San Francisco, CA         CA                6   
1                31834  XNA   Fayetteville, AR         AR                5   
2                31453  IAH        Houston, TX         TX               48   
3                30325  DEN         Denver, CO         CO                8   
4                30325  DEN         Denver, CO         CO                8   

  Dest_State_Name  CRS_Dep_Time  Dep_Time  Dep_Delay  Dep_Delay_Minutes  \
0      California          1400    1353.0       -7.0                0.0  

Fortunately, this raw data comes with a Duplicate flag column.
I will check the counts for this column to check for any values of 'Y' or Yes.
To be certain, I will also run a duplicate check with the .duplicated command.

In [13]:
df['Duplicate'].value_counts()

Duplicate
N    8091684
Name: count, dtype: int64

In [14]:
# So far so good, now to check for duplicates in the entire dataframe by creating a subset of duplicated rows

# Create a DataFrame containing only the duplicate rows
dups = df[df.duplicated(keep=False)]

# Display the 'dups' DataFrame
print(dups)

Empty DataFrame
Columns: [Year, Quarter, Month, Day_of_Month, Day_of_Week, Flight_Date, Marketing_Airline_Network, DOT_ID_Marketing_Airline, IATA_Code_Marketing_Airline, Flight_Number_Marketing_Airline, IATA_Code_Originally_Scheduled_Code_Share_Airline, Operating_Airline , DOT_ID_Operating_Airline, IATA_Code_Operating_Airline, Tail_Number, Flight_Number_Operating_Airline, Origin_Airport_ID, Origin_Airport_Seq_ID, Origin_City_Market_ID, Origin, Origin_City_Name, Origin_State, Origin_State_Fips, Origin_State_Name, Dest_Airport_ID, Dest_Airport_Seq_ID, Dest_City_Market_ID, Dest, Dest_City_Name, Dest_State, Dest_State_Fips, Dest_State_Name, CRS_Dep_Time, Dep_Time, Dep_Delay, Dep_Delay_Minutes, Dep_Del_15, Departure_Delay_Groups, Dep_Time_Blk, Taxi_Out, Wheels_Off, Wheels_On, Taxi_In, CRS_Arr_Time, Arr_Time, Arr_Delay, Arr_Delay_Minutes, Arr_Del_15, Arrival_Delay_Groups, Arr_Time_Blk, Cancelled, Cancellation_Code, Diverted, CRS_Elapsed_Time, Actual_Elapsed_Time, Air_Time, Flights, Distance,

In [15]:
# No dups!

Now I will address missing and/or null values.

In [16]:
# Find the count of null values in each column
null_counts = df.isnull().sum()

In [17]:
# Display the count of null values
print(null_counts)

Year                        0
Quarter                     0
Month                       0
Day_of_Month                0
Day_of_Week                 0
                       ...   
First_Dep_Time        8034702
Total_Add_G_Time      8034705
Longest_Add_G_Time    8034703
Div_Distance          8070905
Duplicate                   0
Length: 69, dtype: int64


In [18]:
# The IATA_Code_Originally_Scheduled_Code_Share column is nearly completely null
# Given that the Bureau of Transportation Statistics (BTS.gov) stated that this column shouldn't be 
# used for analysis, and it is completely null for 2019, I will be dropping it and keeping the Unique_Code
# column as suggested by BTS.gov
# Here is the language from BTS.gov:
# ""Code assigned by IATA and commonly used to identify a carrier.""
#""As the same code may have been assigned to different carriers over time, the code is not always unique.""
#""For analysis, use the Unique Carrier Code.""

# Drop the 'IATA_Code_Originally_Scheduled_Code_Share_Airline' column
df = df.drop(columns=['IATA_Code_Originally_Scheduled_Code_Share_Airline'])

In [19]:
df.shape

(8091684, 68)

There are a significant amount of null values for Dep_Delay, Taxi_Out, Taxi_In, Arr_Time.
I believe these are null values because the flights were cancelled.
I will check value counts for the Cancelled column. Value counts for '1' should be close if not the same to the values missing.

In [20]:
df['Cancelled'].value_counts()

Cancelled
0.0    7938055
1.0     153629
Name: count, dtype: int64

153,629 flights were cancelled in 2019.

This is fairly close to the missing values in:

Dep_Delay                                             147918
Dep_Delay_Minutes                                     147918
Dep_Del_15                                            147918
Departure_Delay_Groups                                147
Taxi_Out                                              152422
Wheels_Off                                            152422
Wheels_On                                             156917
Taxi_In                                               156
Arr_Time                                              156916
Arr_Delay                                             174420
Arr_Delay_Minutes                                     174420
Arr_Del_15                                            174420
Arrival_Delay_Groups                                  17

The missing values in:

Cancellation_Code                                    7938055 -- are accounted for because it is the same as the total Flights (8091684) minus the cancelled flights (153,629) - which is 7938055. Therefore 7938055 flights have no cancellation code because they were not cancelled.

After careful consideration, I believe it is best to create a cancelled flights flag in order to prevent confusion in future analysis for delays v. cancellation. 4420917918

In [21]:
# Identify rows where flights were cancelled

cancelled_flights = df['Cancelled'] == 1

In [22]:
# Create a column to indicate if the data is missing due to cancellation

df['Missing_Due_To_Cancellation'] = cancelled_flights

Now I will address columns containing mixed datatypes.

In [23]:
# Set the display option to show all rows
pd.set_option('display.max_rows', None)

In [24]:
df.dtypes

Year                                 int64
Quarter                              int64
Month                                int64
Day_of_Month                         int64
Day_of_Week                          int64
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int64
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int64
Operating_Airline                   object
DOT_ID_Operating_Airline             int64
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int64
Origin_Airport_ID                    int64
Origin_Airport_Seq_ID                int64
Origin_City_Market_ID                int64
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                    int64
Origin_State_Name                   object
Dest_Airpor

In [25]:
# I will change many of these datatypes to reduce the amount of memory needed by reducing int64 to int8, 16, or 32 where possible.
# Additionally, I will change the Flight_Date column to a date datatype, instead of an object.
# Lets change Flight_Date first

# Ensure 'Flight_Date' column is in datetime format
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'], errors='coerce')

In [26]:
# Extract only the date part (since we don't have time in the same column)
df['Flight_Date'] = df['Flight_Date'].dt.date

In [27]:
# Verify the result
print(df.head())
print(df.dtypes)

   Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2019        1      1             4            5  2019-01-04   
1  2019        1      1             4            5  2019-01-04   
2  2019        1      1             4            5  2019-01-04   
3  2019        1      1             4            5  2019-01-04   
4  2019        1      1             4            5  2019-01-04   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        UA                     19977   
1                        UA                     19977   
2                        UA                     19977   
3                        UA                     19977   
4                        UA                     19977   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          UA                             5657   
1                          UA                             5658   
2                          UA                             5658

In [28]:
# Change datatypes to maximize available memory

df = df.astype({
    'Year': 'int16',
    'Quarter': 'int16',
    'Month': 'int16',
    'Day_of_Month': 'int16',
    'Day_of_Week': 'int16',
    'DOT_ID_Marketing_Airline': 'int16',
    'Flight_Number_Marketing_Airline': 'int16',
    'DOT_ID_Operating_Airline': 'int16',
    'Flight_Number_Operating_Airline': 'int16',
    'Origin_Airport_ID': 'int16',
    'Origin_Airport_Seq_ID': 'int16',
    'Origin_City_Market_ID': 'int16',
    'Origin_State_Fips': 'int8',
    'Dest_Airport_ID': 'int16',
    'Dest_Airport_Seq_ID': 'int16',
    'Dest_City_Market_ID': 'int16',
    'Dest_State_Fips': 'int16',
    'CRS_Dep_Time': 'int16'
})

In [29]:
df.dtypes

Year                                 int16
Quarter                              int16
Month                                int16
Day_of_Month                         int16
Day_of_Week                          int16
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int16
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int16
Operating_Airline                   object
DOT_ID_Operating_Airline             int16
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int16
Origin_Airport_ID                    int16
Origin_Airport_Seq_ID                int16
Origin_City_Market_ID                int16
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                     int8
Origin_State_Name                   object
Dest_Airpor

Given the amount of missing values in First_Dep_Time, Total_Add_G_Time, Longest_Add_G_Time, and Div_Distance, these columns will not be included in this analysis, as the focus is primarily Airline performance as it relates to Delays and Cancellations, not Diverted Flights. 

In [30]:
# Drop the additional irrelevant columns

df = df.drop(columns=['First_Dep_Time', 'Total_Add_G_Time', 'Longest_Add_G_Time', 'Div_Distance'])

In [31]:
df.shape

(8091684, 65)

In [32]:
df.head()

Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2019        1      1             4            5  2019-01-04   
1  2019        1      1             4            5  2019-01-04   
2  2019        1      1             4            5  2019-01-04   
3  2019        1      1             4            5  2019-01-04   
4  2019        1      1             4            5  2019-01-04   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        UA                     19977   
1                        UA                     19977   
2                        UA                     19977   
3                        UA                     19977   
4                        UA                     19977   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          UA                             5657   
1                          UA                             5658   
2                          UA                             5658   
3                          UA                             5659   
4                          UA                             5660   

  Operating_Airline   DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                 OO                     20304                          OO   
1                 OO                     20304                          OO   
2                 OO                     20304                          OO   
3                 OO                     20304                          OO   
4                 OO                     20304                          OO   

  Tail_Number  Flight_Number_Operating_Airline  Origin_Airport_ID  \
0      N945SW                             5657              14698   
1      N932SW                             5658              12266   
2      N932SW                             5658              14783   
3      N916SW                             5659              12389   
4      N107SY                             5660              13342   

   Origin_Airport_Seq_ID  Origin_City_Market_ID Origin     Origin_City_Name  \
0                  28010                 -31300    SBP  San Luis Obispo, CA   
1                 -18581                  31453    IAH          Houston, TX   
2                 -29026                 -30753    SGF      Springfield, MO   
3                  -6282                  32389    ISN        Williston, ND   
4                  23487                 -32194    MKE        Milwaukee, WI   

  Origin_State  Origin_State_Fips Origin_State_Name  Dest_Airport_ID  \
0           CA                  6        California            14771   
1           TX                 48             Texas            15919   
2           MO                 29          Missouri            12266   
3           ND                 38      North Dakota            11292   
4           WI                 55         Wisconsin            11292   

   Dest_Airport_Seq_ID  Dest_City_Market_ID Dest     Dest_City_Name  \
0               -30224                32457  SFO  San Francisco, CA   
1                19040                31834  XNA   Fayetteville, AR   
2               -18581                31453  IAH        Houston, TX   
3                15090                30325  DEN         Denver, CO   
4                15090                30325  DEN         Denver, CO   

  Dest_State  Dest_State_Fips Dest_State_Name  CRS_Dep_Time  Dep_Time  \
0         CA                6      California          1400    1353.0   
1         AR                5        Arkansas           935     930.0   
2         TX               48           Texas           643     637.0   
3         CO                8        Colorado          1335    1314.0   
4         CO                8        Colorado           836     826.0   

   Dep_Delay  Dep_Delay_Minutes  Dep_Del_15  Departure_Delay_Groups  \
0       -7.0                0.0         0.0                    -1.0   
1       -5.0                0.0         0.0                    -1.

In [33]:
# Save Clean 2019 Data

df.to_csv(os.path.join(path, 'clean_flights2019.csv'))